docs/proposal/20210311-single-controller-multitenancy.md

```diff
@@ -0,0 +1,321 @@
+
+#### Story 3 - Legacy behaviour
+
+Erin is an engineer in a smaller less strict organization. They use a single vCenter and keep their management cluster up-to-date. Erin can create new vSphere clusters while omitting the `vSphereCluster.Spec.IdentityRef` field. The CAPV controller will use the credentials it was given at initialization to create new vSphere clusters.
+
+### Requirements
+
+#### Functional Requirements
+
+##### FR1. CAPV MUST support credentials provided through a `VsphereClusterAccount` and referenced by `VsphereCluster.IdentityRef` field
+
+##### FR2. CAPV MUST support credentials provided through a `Secret` and referenced by `VsphereCluster.IdentityRef` field
+
+##### FR3. CAPV MUST support static credentials
+
+##### FR4. CAPV MUST support clusterctl move scenarios
+
+##### FR5. CAPV MUST prevent privilege escalation allowing users to create clusters in accounts they should not be able to
+
+#### Non-Functional Requirements
+
+##### NFR1. Unit tests MUST exist for cluster and machine controllers that utilize the credentials
+
+##### NFR2. e2e tests MUST exist for multi account scenarios
+
+### Implementation Details/Notes/Constraints
+
+#### Current State
+
+CAPV currently uses a session manager to create and cache sessions. Sessions are created (or retrieved from the cache) during cluster and vSphere VM reconcile loops.
+vSphere cluster uses the session as a sanity check to ensure connectivity. vSphere VM uses the session for VM lifecycle tasks on vcenter and the session is stored in a VMContext.
+
+```go
+type VMContext struct {
+    *ControllerContext
+    VSphereVM   *v1alpha3.VSphereVM
+    PatchHelper *patch.Helper
+    Logger      logr.Logger
+    Session     *session.Session
+}
+```
+
+#### Proposed Changes
+
+The proposed changes below allow users to specify the account to use for clusters during runtime. The credentials can be provided by two approaches:
+
+- Referencing a cluster scoped `VsphereClusterAccount` to be used for credentials
+- Referencing a `Secret` that contains the credentials
+
+Changed Resources
+
+- `VsphereCluster`
+
+New Resources
+
+- cluster-scoped `VsphereClusterAccount` represents the vCenter account to use for reconciliation. This type should also contain references to namespaces that are allowed to use the account.
+
+Changes to VsphereCluster
+A new field is added to the VsphereClusterSpec to reference the `VsphereClusterAccount`. We intend to use `corev1.TypedLocalObjectReference` in order to ensure the only objects that can be referenced are either in the same namespace or are scoped globally.
+
+```go
+// VsphereClusterAccount is the account to be used for vcenter actions
+type VsphereClusterAccount struct {
```

Comment:
I know that naming is being discussed with other providers, so feel to ignore…
Given that this type is referenced by `identityRef`, should this be `VsphereClusterIdentity` (or simply `VSphereIdentity`)?